File: Java/Remove Nth Node From End of List.java
// 0ms, 100.00%; 40MB, 5.06%
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public static ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode pre = head, post = head;
        for (int i = 0; i < n; ++i) {
            post = post.next;
        }
        if (post == null) {
            return head.next;
        }
        while (post.next != null) {
            pre = pre.next;
            post = post.next;
        }
        pre.next = pre.next.next;
        return head;
    }
}

File: Go/Remove Nth Node From End of List.go
// 4ms, 17.45%; 2.4MB, 5.06%
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	pre, post := head, head
	for i := 0; i < n; i++ {
		post = post.Next
	}
	if post == nil {
		return head.Next
	}
	for post.Next != nil {
		pre, post = pre.Next, post.Next
	}
	pre.Next = pre.Next.Next
	return head
}
